implementations/rust/ockam/ockam_core/src/routing/message/local_message.rs

```diff
@@ -1,116 +1,210 @@
+
+    /// Return the message return route
+    pub fn return_route(&self) -> Route {
+        self.return_route.clone()
+    }
+
+    /// Return a reference to the message return route
+    pub fn return_route_ref(&self) -> &Route {
+        &self.return_route
+    }
+
+    /// Set the message return route
+    pub fn set_return_route(mut self, route: Route) -> Self {
+        self.return_route = route;
+        self
+    }
+
+    /// Prepend an address to the return route
+    pub fn push_front_return_route(mut self, address: &Address) -> Self {
+        self.return_route.modify().prepend(address.clone());
+        self
+    }
+
+    /// Prepend a route to the return route
+    pub fn prepend_front_return_route(mut self, route: &Route) -> Self {
+        self.return_route.modify().prepend_route(route.clone());
+        self
+    }
+
+    /// Remove the first address on the onward route and push another address on the return route
+    pub fn forward(self, address: &Address) -> Result<Self> {
+        Ok(self
+            .pop_front_onward_route()?
+            .push_front_return_route(address))
+    }
+
+    /// Return the message payload
+    pub fn payload(&self) -> Vec<u8> {
+        self.payload.clone()
+    }
+
+    /// Return a reference to the message payload
+    pub fn payload_ref(&self) -> &[u8] {
+        &self.payload
+    }
+
+    /// Return a mutable reference to the message payload
+    pub fn payload_mut(&mut self) -> &mut Vec<u8> {
+        &mut self.payload
+    }
+
+    /// Prepend an address to the return route
+    pub fn set_payload(mut self, payload: Vec<u8>) -> Self {
+        self.payload = payload;
+        self
+    }
+
+    /// Return the message local info
+    pub fn local_info(&self) -> Vec<LocalInfo> {
+        self.local_info.clone()
+    }
+
+    /// Return a reference to the message local info
+    pub fn local_info_ref(&self) -> &[LocalInfo] {
         &self.local_info
     }
-    /// Dissolve
-    pub fn dissolve(self) -> (TransportMessage, Vec<LocalInfo>) {
-        (self.transport_message, self.local_info)
+
+    /// Return a mutable reference to the message local info
+    pub fn local_info_mut(&mut self) -> &mut Vec<LocalInfo> {
+        &mut self.local_info
     }
-}
 
-impl LocalMessage {
-    /// Append a new [`LocalInfo`] entry.
-    pub fn append_local_info(&mut self, local_info: LocalInfo) {
-        self.local_info.push(local_info)
+    /// Clear all [`LocalInfo`] entries
+    pub fn clear_local_info(&mut self) {
+        self.local_info.clear()
     }
 
-    /// Replace all [`LocalInfo`] entries matching the type identifier
-    /// of the given `LocalInfo` with itself.
-    pub fn replace_local_info(&mut self, local_info: LocalInfo) {
-        self.clear_local_info(local_info.type_identifier());
-        self.local_info.push(local_info)
+    /// Create a [`LocalMessage`] from a decoded [`TransportMessage`]
+    pub fn from_transport_message(transport_message: TransportMessage) -> LocalMessage {
```

Comment:
I'm a bit on the fence with using `From` when there's only one conversion in one direction and / or the other. Because with implicit conversions it's not immediately apparent what are things transformed into. It seems to me that it's better left when there are lots of conversions `X -> Y` and then having many functions `from_x1`, `from_x2`, `from_x3`,... is not that great.